File: ios/Classes/NuveiPaymentWrapperPlugin.swift
import Flutter
import UIKit
import NuveiSimplyConnectSDK

public class NuveiPaymentWrapperPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "nuvei_payment_wrapper", binaryMessenger: registrar.messenger())
    let instance = NuveiPaymentWrapperPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let viewController: UIViewController = (UIApplication.shared.delegate?.window??.rootViewController)!;

    // Method Channel       
    switch call.method {
      case "setup":
        guard let args = call.arguments as? [String : Any] else {return}
        self.setup(result: result, args: args)
        break
      case "authenticate3d":
        guard let args = call.arguments as? [String : Any] else {return}
        self.authenticate3d(result: result, args: args, controller: viewController)
        break
      default:
        result(FlutterMethodNotImplemented)
        break
    }
  }
    
    /* Set environment */
     private func setup(result: FlutterResult, args: [String : Any]) {
       let environment = args["environment"] as! String
       switch environment {
            case PackageEnvironment.stating:
               NuveiSimplyConnect.setup(environment: NuveiSimplyConnect.Environment.integration)
               break
            default:
               NuveiSimplyConnect.setup(environment: NuveiSimplyConnect.Environment.production)
               break
       }
         
        result(true)
     }
    
     /* Authenticate 3D */
     private func authenticate3d(result: @escaping FlutterResult, args: [String : Any], controller: UIViewController) {
       let sessionToken = args["sessionToken"] as! String
       let merchantId = args["merchantId"] as! String
       let merchantSiteId = args["merchantSiteId"] as! String
       let currency = args["currency"] as! String
       let amount = args["amount"] as! String
       let cardHolderName = args["cardHolderName"] as! String
       let cardNumber = args["cardNumber"] as! String
       let cvv = args["cvv"] as! String
       let monthExpiry = args["monthExpiry"] as! String
       let yearExpiry = args["yearExpiry"] as! String
      
       let paymentOption = try! NVPaymentOption(
         card: NVCardDetails(
           cardNumber: cardNumber,
           cardHolderName: cardHolderName,
           cvv: cvv,
           expirationMonth: monthExpiry,
           expirationYear: yearExpiry
         )
       )
       let input = NVInput(
         sessionToken: sessionToken,
         merchantId: merchantId,
         merchantSiteId: merchantSiteId,
         currency: currency,
         amount: amount,
         paymentOption: paymentOption
       );
      
       // The viewController is used to present the challenge view controller(s) from
       NuveiSimplyConnect.authenticate3d(uiOwner: controller, input: input) { (output: NVAuthenticate3dOutput) in
         let authenticate3dResponse:Authenticate3dResponse = Authenticate3dResponse(
           cavv: output.cavv,
           eci: output.eci,
           dsTransID: output.dsTransID,
           ccTempToken: output.ccTempToken,
           transactionId: output.transactionId,
           result: output.result.rawValue.uppercased(),
           transactionStatus: output.rawResult?["transactionStatus"] as? String ?? "",
           errorDescription: output.errorDescription,
           errCode: output.errorCode,
           status: output.rawResult?["status"] as? String ?? ""
         )
         let authenticate3dResponseToJson = self.convertToJson(data: authenticate3dResponse)

         result(authenticate3dResponseToJson)
       }        
     }
           
    // utils function
    private func convertToJson<T: Encodable>(data: T) -> String {
      let encoder = JSONEncoder()
      encoder.outputFormatting = .prettyPrinted

      let data = try! encoder.encode(data)
      print(String(data: data, encoding: .utf8)!)
      
      return String(data: data, encoding: .utf8)!
    }  
}

enum PackageEnvironment {
  static let stating = "STAGING"
  static let production = "PRODUCTION"
}

struct Authenticate3dResponse: Codable {
  var cavv: String?
  var eci: String?
  var dsTransID: String?
  var ccTempToken: String?
  var transactionId: String?
  var result: String?
  var transactionStatus: String?
  var errorDescription: String?
  var errCode: Int?
  var status: String?
}

File: android/src/main/kotlin/com/exnodes/nuvei_payment_wrapper/NuveiPaymentWrapperPlugin.kt
package com.exnodes.nuvei_payment_wrapper
import android.app.Activity
import android.content.Context
import com.google.gson.Gson
import android.util.Log
import androidx.annotation.NonNull
import com.nuvei.sdk.Callback

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import com.nuvei.sdk.NuveiSimplyConnect
import com.nuvei.sdk.model.*
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/** NuveiPaymentWrapperPlugin */
class NuveiPaymentWrapperPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var context: Context
  private lateinit var activity: Activity
  private val gson = Gson()

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "nuvei_payment_wrapper")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "setup" -> {
        setup(result, call)
      }
      "authenticate3d" -> {
        authenticate3d(result, call)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun setup(result: MethodChannel.Result, call: MethodCall) {
    // Set environment
    val environment: String = call.argument("environment")!!
    when (environment) {
      PackageEnvironment.stating -> {
        NuveiSimplyConnect.setup(
          environment = NuveiSimplyConnect.Environment.STAGING,
        );
      }
      else -> {
        NuveiSimplyConnect.setup(
          environment = NuveiSimplyConnect.Environment.PROD,
        );
      }
    }

    result.success(true)
  }

  private fun authenticate3d(result: MethodChannel.Result, call: MethodCall) {
    val sessionToken: String = call.argument("sessionToken")!!
    val merchantId: String = call.argument("merchantId")!!
    val merchantSiteId: String = call.argument("merchantSiteId")!!
    val currency: String = call.argument("currency")!!
    val amount: String = call.argument("amount")!!
    val cardHolderName: String = call.argument("cardHolderName")!!
    val cardNumber: String = call.argument("cardNumber")!!
    val cvv: String = call.argument("cvv")!!
    val monthExpiry: String = call.argument("monthExpiry")!!
    val yearExpiry: String = call.argument("yearExpiry")!!

    val paymentOption = PaymentOption(
      card = CardDetails(
        cardNumber,
        cardHolderName,
        cvv,
        monthExpiry,
        yearExpiry,
      ),
    )

    val input = NVPayment(
      sessionToken,
      merchantId,
      merchantSiteId,
      currency,
      amount,
      paymentOption = paymentOption,
    )

    NuveiSimplyConnect.authenticate3d(
      activity = activity,
      input = input,
      callback = object : Callback<NVAuthenticate3dOutput> {
        override fun onComplete(response: NVAuthenticate3dOutput) {
           writeToLog(gson.toJson(response))
          val authenticate3dResponse = Authenticate3dResponse(
            cavv = response.cavv,
            eci =  response.eci,
            dsTransID = response.dsTransID,
            ccTempToken =  response.ccTempToken,
            transactionId = response.transactionId,
            result = response.result.uppercase(),
            transactionStatus = response.rawResult?.get("transactionStatus") as String?,
            errorDescription = response.errorDescription,
            errCode = response.errorCode,
            status = response.rawResult?.get("status") as String?
          )
          val authenticate3dResponseToJson = gson.toJson(authenticate3dResponse)
          result.success(authenticate3dResponseToJson)
        }
      }
    )
  }

  private fun writeToLog(log: String) {
    Log.d("print", log)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity;
  }

  override fun onDetachedFromActivityForConfigChanges() {
    TODO("Not yet implemented")
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    TODO("Not yet implemented")
  }

  override fun onDetachedFromActivity() {
    TODO("Not yet implemented")
  }
}

class PackageEnvironment {
  companion object {
    const val stating = "STAGING"
    const val production = "PRODUCTION"
  }
}

data class Authenticate3dResponse(
  val cavv: String?,
  val eci: String?,
  val dsTransID: String?,
  val ccTempToken: String?,
  val transactionId: String?,
  val result: String?,
  val transactionStatus: String?,
  val errorDescription: String?,
  val errCode: Int?,
  val status: String?,
)
